content/2019-03-18-beniget.rst

```diff
@@ -0,0 +1,211 @@
+...
+...     def __init__(self, module_node):
+...         # initialize def-use chains
+...         self.chains = beniget.DefUseChains()
+...         self.chains.visit(module_node)
+...         self.users = set()  # users of local definitions
+...         self.captured = set()  # identifiers that don't belong to local users
+...
+...     def visit_FunctionDef(self, node):
+...         # initialize the set of node using a local variable
+...         for def_ in self.chains.locals[node]:
+...             self.users.update(use.node for use in def_.users())
+...         self.generic_visit(node)
+...
+...     def visit_Name(self, node):
+...         # register load of identifiers not locally defined
+...         if isinstance(node.ctx, ast.Load):
+...             if node not in self.users:
+...                 self.captured.add(node.id)
+
+>>> code = 'def foo(x):\n def bar(): return x\n return bar'
+>>> module = ast.parse(code)
+>>> inner_function = module.body[0].body[0]
+>>> capture = Capture(module)
+>>> capture.visit(inner_function)
+>>> list(capture.captured)
+['x']
+
+Detecting ``NameError``
+-----------------------
+
+Any ``USE`` without ``DEF`` is probably (and no, not **certainly**) an error. But even if there's an associated ``DEF``, it could be an error; Consider the following:
+
+.. code-block:: python
+
+    from random import random
+    if random() > 0.5:
+        a = 1
+    print(a)
+
+There's a chance that ``a`` is unbound while executing ``print(a)``. It would be possible to combine ``beniget`` with a dummy code transformation to detect this issue by generating dummy top-level definitions and checking if they have any ``USE``:
+
+
+.. code-block:: python
+
+    a = random = None  # if any of those have a USE, then we have a potential issue
+    from random import random
+    if random() > 0.5:
+        a = 1
+    print(a)
+
+Limitations
+-----------
+
+It's Python. So introspection and lazy binding are a pain for *any* static analysis tool. There's nothing we can do against
+
+- `eval <https://docs.python.org/3/library/functions.html#eval>`_ and `exec <https://docs.python.org/3/library/functions.html#exec>`_
+- star import ``from some_module import *`` even if ``beniget``, in a very conservative way, assumes that such import can define any identifier, which means it's likely to have a lot of ``USE``!
+- assigning to ``globals()`` or ``locals()``
+
+And plenty of other similar stuff. I can't blame you for using these features, that's part of `Python nature <https://fr.wikipedia.org/wiki/Le_Scorpion_et_la_Grenouille>`_
+
+Also, note that ``beniget`` analysis is not data dependant, so ``if 1: a = 1`` does **not** unconditionally defines ``a``!
+
+Installation and stuff
+----------------------
+
+``beinget`` is available on `PyPI <https://pypi.org/project/beniget/>`_ and `GitHub <https://github.com/serge-sans-paille/beniget/>`_ under `BSD 3-clause <https://github.com/serge-sans-paille/beniget/blob/master/LICENSE>`_.
```

Comment:
beniget

You have picked a very challenging name for this project! :smile: For the above reason, and because there is another completely unrelated project from [intel/beignet](https://github.com/intel/beignet).